### DISCUSSION TITLE
soft_hard_data_fusion

### DISCUSSION THREAD

**Discussion by SohSalari** [dstl/Stone-Soup#423] [Ideas]
I would like to implement a modified version of GM-PHD filter, which is able to accept and process (fuse) both soft (human-generated) measurements and hard (sensor-provided) measurements. 

Let us assume that soft measurements can be modeled as the summation of weighted Gaussian components, in which each component consists of 3 parts:  1) weight (probability), 2) state_vector, and 3) Covariance_Matrix.  For example: soft_measurement = w_1 * N(m_1, Q _1) + w_2 N(m_2, Q_2) 

Is there any way to generate soft measurements, as the summation of weighted Gaussian components, using available "Detection Types" of Stone-Soup or need to define a new class for this purpose?

**Comment by sdhiscocks** [comment]
This sounds really interesting @SohSalari.

So we don't have anything like this yet, but could easily use our exiting components to construct something.

```python
import numpy as np

from stonesoup.types.detection import Detection
from stonesoup.types.state import WeightedGaussianState
from stonesoup.types.mixture import GaussianMixture
from stonesoup.types.array import StateVector, CovarianceMatrix

class GaussianMixtureDetection(Detection, GaussianMixture):
    state_vector = None  # Remove state_vector inherited from Detection

mean_1 = StateVector([0, 0])
mean_2 = StateVector([1, 2])
covar_1 = CovarianceMatrix(np.diag([1, 1]))
covar_2 = CovarianceMatrix(np.diag([2, 3]))

soft_measurement = GaussianMixtureDetection([
    WeightedGaussianState(mean_1, covar_1, weight=0.4),
    WeightedGaussianState(mean_2, covar_2, weight=0.6)])
```

Also looked at adding some methods that may be useful in 6ce0745, which you may want to consider.

**Comment by SohSalari** [comment]
Thanks @sdhiscocks.

I would like to implement a filter within the Stone Soup, which is able to fuse provided measurements in different coordinates. For example, consider the case where human (soft sensor) measures in Cartesian coordinate, while radar (hard sensor) provides measurements in Poolar coordinate. How can we use the same (modified gm-phd) filter to fuse these Cartesian-Poolar measurements? Is there any way to use Stone-Soup's existing components to handle this case?

**Comment by sdhiscocks** [comment]
So coordinate conversion is done by the `MeasurementModel`s. You'll see an example of this in the [EKF tutorial](https://stonesoup.readthedocs.io/en/latest/auto_tutorials/02_ExtendedKalmanFilterTutorial.html#sphx-glr-auto-tutorials-02-extendedkalmanfiltertutorial-py), as the track is in Cartesian.

The `Detection` type allows you to assign a model to the `measurement_model` property, which then the other components can use to do conversions. The [multi-sensor tutorial](https://stonesoup.readthedocs.io/en/latest/auto_examples/Moving_Platform_Simulation.html#sphx-glr-auto-examples-moving-platform-simulation-py) has an example of this, with radar and bearing only sensor (e.g. optical sensor). May be easier to see by looking at a [radar's measure method](https://github.com/dstl/Stone-Soup/blob/master/stonesoup/sensor/radar/radar.py#L225-L255).

**Comment by SohSalari** [comment]
Thanks @sdhiscocks.

How should we set up the PHDUpdater to be able to employ the same **gm-phd** filter for processing the provided measurements by Sensors **A** an **B**? 

1- If sensor (**A**) measures in the Cartesian coordinates (Linear measurement model), we can simply set the following PHDUpdater:
updater = PHDUpdater(
         **_kalman_updater = KalmanUpdater(measurement_model = None_**),
         clutter_spatial_density,
         prob_detection,
         prob_survival)

2- If sensor (**B**) measures in polar coordinates (let assume a Highly Non-Linear measurement model when EKF is not a good option), then we need to set the following PHDUpdater:
updater = PHDUpdater(
         **_unscented_updater = UnscentedKalmanUpdater(measurement_model = None_**),
         clutter_spatial_density,
         prob_detection,
         prob_survival)

We don’t add a measurement model to the updater, since each sensor adds its measurement model to the detection itgenerate. 
But, how can we set the PHDUpdater when we want to use the same filter to process measrements of sensors **A** and **B** ?

**Comment by sdhiscocks** [comment]
So currently there isn't an easy way to do this, as you can only have one updater. The `UnscentedKalmanUpdater` will work with a Linear model, but obviously isn't optimal.

You could create a updater which just wraps existing ones, calling them off depending on type of model. Something like this (I've not tested this):
```python
class UpdaterWrapper():
    linear_updater = KalmanUpdater(None)
    non_linear_updater = UnscentedKalmanUpdater(None)
    
    @staticmethod
    def _pick_updater(measurement_model):
        if isinstance(measurement_model, LinearModel):
            return self.linear_updater
        else:
            return self.non_linear_updater
    
    def predict_measurement(self, predicted_state, measurement_model, **kwargs):
        updater = self._pick_updater(measurement_model)
        return updater.predict_measurement(predicted_state, measurement_model, **kwargs)
    
    def update(self, hypothesis, **kwargs):
        updater = self._pick_updater(hypothesis.measurement.measurement_model)
        return updater.update(hypothesis, **kwargs)
```

```python
updater = PHDUpdater(
    UpdaterWrapper(),
    clutter_spatial_density,
    prob_detection,
    prob_survival)
```
